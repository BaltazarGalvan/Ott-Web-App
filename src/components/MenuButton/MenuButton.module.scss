@use '../../styles/variables';
@use '../../styles/theme';
@use '../../styles/mixins/responsive';
@use '../../styles/mixins/utils';

//
// Navigation Button
// --------------------------------

.menuButton {
  position: relative;
  display: inline-flex;
  justify-content: inherit;
  align-items: center;
  min-height: 44px;
  padding: 0 variables.$base-spacing;
  color: variables.$white;
  font-family: theme.$body-alt-font-family;
  font-weight: 700;
  font-size: 18px;
  text-align: center;
  text-decoration: none;
  background: transparent;
  cursor: pointer;
  opacity: 0.7;
  transition: background 0.1s ease;
  outline-color: var(--highlight-color, white);

  @media (hover: hover) and (pointer: fine) {
    &:hover,
    &:active {
      background: rgba(variables.$white, 0.08);
      opacity: 1;
    }
  }
}

.active {
  background: rgba(variables.$white, 0.08);
  opacity: 1;
}
