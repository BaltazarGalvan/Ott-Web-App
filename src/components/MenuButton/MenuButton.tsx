import React from 'react';
import { NavLink } from 'react-router-dom';
import classNames from 'classnames';

import styles from './MenuButton.module.scss';

type Props = {
  label?: string;
  to?: string;
  onClick?: () => void;
  active?: boolean;
};

const MenuButton: React.FC<Props> = ({ label, to, onClick, active = false }: Props) => {
  if (to) {
    return (
      <NavLink className={styles.menuButton} activeClassName={styles.active} to={to} exact>
        <span className={styles.label}>{label}</span>
      </NavLink>
    );
  }

  return (
    <div className={classNames(styles.menuButton, { [styles.active]: active })} onClick={onClick}>
      <span className={styles.label}>{label}</span>
    </div>
  );
};

export default MenuButton;
