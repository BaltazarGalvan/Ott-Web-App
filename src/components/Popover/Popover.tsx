import React, { ReactNode, useRef } from 'react';
import classNames from 'classnames';

import DetectOutsideClick from '../DetectOutsideClick/DetectOutsideClick';
import Slide from '../Animation/Slide/Slide';

import styles from './Popover.module.scss';

type Props = {
  children: ReactNode;
  isOpen: boolean;
  onClose: () => void;
};

const Popover: React.FC<Props> = ({ children, isOpen, onClose }: Props) => {
  const popoverRef = useRef<HTMLDivElement>(null);

  return isOpen ? (
    <DetectOutsideClick isActive={isOpen} callback={onClose} el={popoverRef}>
      <Slide open={isOpen} duration={300} onCloseAnimationEnd={() => onClose()} fromRight>
        <div ref={popoverRef} className={classNames(styles.popover)}>
          {children}
        </div>
      </Slide>
    </DetectOutsideClick>
  ) : null;
};

export default Popover;
