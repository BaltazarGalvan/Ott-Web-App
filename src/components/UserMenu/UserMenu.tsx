import React, { useCallback } from 'react';
import { useNavigate } from 'react-router-dom';
import { useTranslation } from 'react-i18next';
import classNames from 'classnames';

import styles from './UserMenu.module.scss';

import AccountCircle from '#src/icons/AccountCircle';
import Favorite from '#src/icons/Favorite';
import BalanceWallet from '#src/icons/BalanceWallet';
import Exit from '#src/icons/Exit';
import MenuButton from '#components/MenuButton/MenuButton';
import { logout } from '#src/stores/AccountController';
import { useConfigStore } from '#src/stores/ConfigStore';
import LoadingOverlay from '#components/LoadingOverlay/LoadingOverlay';
import Plus from '#src/icons/Plus';
import { useSelectProfile, useListProfiles, unpersistProfile, useProfilesFeatureEnabled } from '#src/hooks/useProfiles';
import ProfileCircle from '#src/icons/ProfileCircle';
import { useProfileStore } from '#src/stores/ProfileStore';

type Props = {
  small?: boolean;
  showPaymentsItem: boolean;
  onClick?: () => void;
};

const UserMenu = ({ showPaymentsItem, small = false, onClick }: Props) => {
  const { t } = useTranslation('user');
  const navigate = useNavigate();
  const { accessModel } = useConfigStore();
  const { profile: currentProfile } = useProfileStore();
  const profilesEnabled = useProfilesFeatureEnabled();

  const { data, isFetching } = useListProfiles();
  const profiles = data?.responseData.collection;

  if (profilesEnabled && !profiles?.length) {
    unpersistProfile();
    navigate('/u/profiles');
  }

  const selectProfile = useSelectProfile();

  const onLogout = useCallback(async () => {
    if (onClick) {
      onClick();
    }

    await logout();
    navigate('/', { replace: true });
  }, [onClick, navigate]);

  return (
    <ul className={styles.menuItems}>
      {accessModel === 'SVOD' && profilesEnabled && (
        <>
          <div className={styles.sectionHeader}>{t('nav.switch_profiles')}</div>
          {selectProfile.isLoading || isFetching ? (
            <LoadingOverlay inline />
          ) : (
            profiles?.map((profile) => (
              <li key={profile.id}>
                <MenuButton
                  active={profile.id === currentProfile?.id}
                  small={small}
                  onClick={() => selectProfile.mutate({ id: profile.id, avatarUrl: profile.avatar_url })}
                  label={profile.name}
                  startIcon={<ProfileCircle src={profile.avatar_url} alt={profile.name} />}
                />
              </li>
            ))
          )}
          {(profiles?.length || 0) < 4 && (
            <li>
              <MenuButton small={small} onClick={() => navigate('/u/profiles/create')} label={t('nav.add_profile')} startIcon={<Plus />} />
            </li>
          )}
          <hr
            className={classNames(styles.divider, {
              [styles.small]: small,
            })}
          />
        </>
      )}
      <div className={styles.sectionHeader}>{t('nav.settings')}</div>
      {profilesEnabled && (
        <li>
          <MenuButton
            small={small}
            onClick={onClick}
            to={`/u/my-profile/${currentProfile?.id ?? ''}`}
            label={t('nav.profile')}
            startIcon={<ProfileCircle src={currentProfile?.avatar_url ?? ''} alt={currentProfile?.name ?? ''} />}
          />
        </li>
      )}
      <li>
        <MenuButton small={small} onClick={onClick} to="/u/my-account" label={t('nav.account')} startIcon={<AccountCircle />} />
      </li>

      <li>
        <MenuButton small={small} onClick={onClick} to="/u/favorites" label={t('nav.favorites')} startIcon={<Favorite />} />
      </li>
      {showPaymentsItem && (
        <li>
          <MenuButton small={small} onClick={onClick} to="/u/payments" label={t('nav.payments')} startIcon={<BalanceWallet />} />
        </li>
      )}
      <hr className={classNames(styles.divider, { [styles.small]: small })} />
      <li>
        <MenuButton small={small} onClick={onLogout} label={t('nav.logout')} startIcon={<Exit />} />
      </li>
    </ul>
  );
};

export default UserMenu;
