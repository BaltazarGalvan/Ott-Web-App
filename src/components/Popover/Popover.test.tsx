import React from 'react';
import { render } from '@testing-library/react';

import MenuButton from '../MenuButton/MenuButton';
import UserMenu from '../UserMenu/UserMenu';

import Popover from './Popover';

describe('<Popover>', () => {
  test('renders and matches snapshot', () => {
    const menuItems = [<MenuButton key="key" label="Home" to="/" />];
    const { container } = render(<Popover popoverButtonIcon={<UserMenu />}>{menuItems}</Popover>);

    expect(container).toMatchSnapshot();
  });
});
