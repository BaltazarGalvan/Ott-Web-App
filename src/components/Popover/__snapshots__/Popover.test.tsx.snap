// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Popover> renders and matches snapshot 1`] = `<div />`;
